### PR TITLE
Add @typePolicy(embeddedFields: String! = "")

### PR DESCRIPTION
**Pull request by martinbonnin**
`@embed` allows "embedding" objects in their parent records. Will be useful for pagination. This schema:

```grahql
type Query @typePolicy(embeddedFields: "hero"){
  hero: Hero! 
}

type Hero @typePolicy(embeddedFields: "friends"){
  friends: [Friend!]! 
}

type Friend {
  name: String!
}
```

Will only produce a single Record. Note that it's opt-in because in the worst case where no object has an id, we could rewrite the whole cache on each query.

### CODE REVIEW

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/CompiledGraphQL.kt [hunk 1]
```diff
@@ -158,17 +158,111 @@ class CustomScalarType(
...
+  constructor(
+      name: String,
+      keyFields: List<String> = emptyList(),
+      implements: List<InterfaceType> = emptyList(),
+  ) : this(name, keyFields, implements, emptyList())
+
+  fun newBuilder(): Builder = Builder(this)
+
+  class Builder(internal val name: String) {
+    internal var keyFields: List<String> = emptyList()
```
Comment (BoD) [inline apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/CompiledGraphQL.kt:182]: Nit: could be `private`?

**Review by BoD — APPROVED** [review]

**Comment by martinbonnin** [comment]
I updated this PR to leverage the existing `@typePolicy`, no only is this more in line with `keyFields` but it also makes possible to extend the existing types while keeping a GraphQL compatible syntax. Let me know what you think

File: apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/Schema.kt [hunk 2]
```diff
@@ -188,6 +193,7 @@ class Schema internal constructor(
     const val NONNULL = "nonnull"
     const val OPTIONAL = "optional"
     const val REQUIRES_OPT_IN = "requiresOptIn"
+    const val EMBED = "embed"
```
Comment (BoD) [inline apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/Schema.kt:196]: Can be deleted

File: kotlin-js-store/yarn.lock [hunk 3]
```diff
@@ -49,11 +49,6 @@ argparse@^2.0.1:
   resolved "https://registry.yarnpkg.com/argparse/-/argparse-2.0.1.tgz#246f50f3ca78a3240f6c997e8a9bd1eac49e4b38"
   integrity sha512-8+9WqebbFzpX9OR+Wa6O29asIogeRMzcGtAINdpMHHyAg10f05aSFVBbcEqGf/PXw1EjAZ+q2/bEBg3DvurK3Q==
 
-async-limiter@~1.0.0:
```
Comment (BoD) [inline kotlin-js-store/yarn.lock:52]: It happened again!

**Review by BoD — APPROVED** [review]

**Comment by BoD** [comment]
> I updated this PR to leverage the existing @typePolicy, no only is this more in line with keyFields but it also makes possible to extend the existing types while keeping a GraphQL compatible syntax. Let me know what you think

Yes, even better, smaller API surface 👍

File: kotlin-js-store/yarn.lock (hunk 3 above)
Comment (martinbonnin) [reply kotlin-js-store/yarn.lock]: Fixed with https://github.com/apollographql/apollo-kotlin/pull/4199

**Pull request merged**